packages/rooks/src/hooks/useFullscreen.ts

```diff
@@ -1,266 +1,226 @@
+    );
+  },
+
+  get fullscreenElement() {
+    if (typeof document === "undefined") return null;
+
+    return (
+      document.fullscreenElement ||
+      document.webkitFullscreenElement ||
+      document.webkitCurrentFullScreenElement ||
+      document.mozFullScreenElement ||
+      document.msFullscreenElement ||
+      null
+    );
+  },
+
+  requestFullscreen(
+    element: Element | null,
+    options?: FullscreenOptions | undefined
+  ): Promise<void> {
+    if (typeof document === "undefined")
+      throw new Error(this.DOM_UNAVAILABLE_ERROR);
+
+    const target = element ?? document.documentElement;
+
+    const method =
+      target.requestFullscreen ||
+      target.webkitRequestFullscreen ||
+      target.webkitRequestFullScreen ||
+      target.mozRequestFullScreen ||
+      target.msRequestFullscreen;
+
+    if (!method) throw new Error(this.FULLSCREEN_UNSUPPORTED_ERROR);
+
+    return method.call(target, options);
+  },
+
+  exitFullscreen(): Promise<void> {
+    if (typeof document === "undefined")
+      throw new Error(this.DOM_UNAVAILABLE_ERROR);
+
+    const method =
+      document.exitFullscreen ||
+      document.webkitExitFullscreen ||
+      document.webkitCancelFullScreen ||
+      document.mozCancelFullScreen ||
+      document.msExitFullscreen;
+
+    if (!method) throw new Error(this.FULLSCREEN_UNSUPPORTED_ERROR);
+
+    return method.call(document);
+  },
+
+  on(eventType: "change" | "error", callback: (event: Event) => void) {
+    const eventName = this.getEventName(eventType);
+    if (!eventName) return;
+
+    document?.addEventListener(eventName, callback);
+  },
+
+  off(eventType: "change" | "error", callback: (event: Event) => void) {
+    const eventName = this.getEventName(eventType);
+    if (!eventName) return;
+
+    document?.removeEventListener(eventName, callback);
+  },
 };
 
-function warnDeprecatedOnChangeAndOnErrorUsage() {
-  warning(
-    false,
-    `Using onChange and onError from the return value is deprecated and 
-    will be removed in the next major version. 
-    Please use it with arguments instead. 
-    For eg: useFullscreen({onChange: function() {}, onError: function(){}})
-  `
-  );
+interface UseFullscreenProps {
+  target?: RefObject<Element>;
+  onChange?: (event: Event) => void;
+  onError?: (event: Event) => void;
+  requestFullScreenOptions?: FullscreenOptions;
 }
 
-/**
- * useFullscreen
- * A hook that helps make the document fullscreen
- */
-function useFullscreen(options: FullScreenOptions = {}): FullscreenApi {
-  const {
-    onChange: onChangeArgument,
-    onError: onErrorArgument,
-    requestFullscreenOptions = {},
-  } = options;
-
-  const fullscreenControls = getFullscreenControls();
-  const [isFullscreen, setIsFullscreen] = useState(
-    Boolean(document[fullscreenControls.fullscreenElement])
-  );
-  const [element, setElement] = useState(
-    document[fullscreenControls.fullscreenElement]
+function useFullscreen(props: UseFullscreenProps = {}) {
```

Comment:
```suggestion
/**
 *
 * useFullscreen hook
 *
 * Gives control to make HTML Elements fullscreen.
 *
 * @param {Element | undefined} props.target The target element to be fullscreen.
 * @param {(event: Event) => void} props.onChange The function to be called when the fullscreen changes.
 * @param {(event: Event) => void} props.onError The function to be called when the fullscreen error occurs.
 * @param {FullscreenOptions} props.requestFullscreenOptions The options to be passed to the requestFullscreen function.
 * @return {Object} returns - The controlls of useFullscreen hook.
 * @return {boolean} returns.isFullscreenAvailable - Whether the fullscreen is available.
 * @return {Element | null} returns.fullscreenElement - The fullscreen element.
 * @return {boolean} returns.isFullscreenEnabled - Whether the fullscreen is enabled.
 * @return {() => Promise<void>} returns.enableFullscreen - The function to enable fullscreen.
 * @return {() => Promise<void>} returns.disableFullscreen - The function to disable fullscreen.
 * @return {() => Promise<void>} returns.toggleFullscreen - The function to toggle fullscreen.
 */
function useFullscreen(props: UseFullscreenProps = {}) {
```

I tried adding JSDoc.

Reply:
I would also add 

@param {Object} props Optional hook configuration

or something like that.

Reply:
I tried it at the beginning, but it will prevent the VSCode hovering window from showing what's inside of the props.

With `@param {Object}`:
<img width="413" alt="スクリーンショット 2022-08-14 1 51 03" src="https://user-images.githubusercontent.com/17402261/184503528-114d01fa-6c21-4a7c-af67-5df957b0e41d.png">

Without it:
<img width="637" alt="スクリーンショット 2022-08-14 1 50 28" src="https://user-images.githubusercontent.com/17402261/184503541-dca82cc0-4670-49e7-b12f-3895e42a7b09.png">

Reply:
I see. Looks good to me